File: controller-2/Core/Src/comm/cobs.c
/**
 * @file cobs.c
 * @brief Implementation COBS. Voir comm/cobs.h pour le pourquoi.
 *
 * L'encodage suit la forme canonique de l'article d'origine, y compris sur le cas limite du
 * groupe plein qui tombe exactement en fin de donnees : aucun octet de code supplementaire
 * n'est emis. Ce detail n'a aucune importance pour le decodeur — les deux formes se decodent
 * identiquement — mais il en a une pour nous : les deux implementations, C et TypeScript,
 * doivent produire exactement les memes octets, sinon les vecteurs de test partages divergent
 * sur ce seul cas et la panne se revele le jour ou une trame fait pile 254 octets sans zero.
 */
#include "comm/cobs.h"

size_t Cobs_Encode(const uint8_t *src, size_t len, uint8_t *dst, size_t dst_cap)
{
  if ((src == NULL) || (dst == NULL)) {
    return 0U;
  }
  if (dst_cap < COBS_MAX_ENCODED(len)) {
    return 0U;
  }

  size_t  out      = 1U;   /* la place du premier octet de code est reservee d'emblee */
  size_t  code_pos = 0U;   /* ou ecrire le code du groupe en cours */
  uint8_t code     = 1U;   /* nombre d'octets du groupe, octet de code compris */

  for (size_t i = 0U; i < len; i++) {
    if (src[i] != 0U) {
      dst[out++] = src[i];
      code++;
    }

    /* Un groupe se ferme sur un zero — que le code remplace — ou parce qu'il est plein. */
    if ((src[i] == 0U) || (code == 0xFFU)) {
      dst[code_pos] = code;
      code     = 1U;
      code_pos = out;

      /* On ne reserve un octet de code pour le groupe suivant que s'il portera quelque
       * chose : soit le zero qu'on vient de consommer avait un successeur, soit il reste
       * des donnees. Sinon le code final ecrit ci-dessous tombe hors de la longueur
       * rendue, et c'est exactement ce qu'on veut. */
      if ((src[i] == 0U) || ((i + 1U) < len)) {
        out++;
      }
    }
  }

  dst[code_pos] = code;
  return out;
}

size_t Cobs_Decode(const uint8_t *src, size_t len, uint8_t *dst, size_t dst_cap)
{
  if ((src == NULL) || (dst == NULL) || (len == 0U)) {
    return 0U;
  }

  size_t in  = 0U;
  size_t out = 0U;

  while (in < len) {
    const uint8_t code = src[in];

    /* Un code nul est impossible dans un flux COBS valide : 0x00 est le delimiteur, il ne
     * doit jamais parvenir jusqu'ici. Le rencontrer signale un flux corrompu. */
    if (code == 0U) {
      return 0U;
    }
    /* Le groupe annonce code-1 octets litteraux : ils doivent tenir dans l'entree. */
    if ((in + (size_t)code) > len) {
      return 0U;
    }
    in++;

    for (uint8_t k = 1U; k < code; k++) {
      if (out >= dst_cap) {
        return 0U;
      }
      dst[out++] = src[in++];
    }

    /* Zero implicite en fin de groupe, sauf si le groupe etait plein (0xFF, aucun zero a
     * restituer) et sauf en toute fin d'entree, ou il n'y a plus rien a separer. */
    if ((code != 0xFFU) && (in < len)) {
      if (out >= dst_cap) {
        return 0U;
      }
      dst[out++] = 0U;
    }
  }

  return out;
}

File: interface/src/shared/cobs.ts
/**
 * COBS — Consistent Overhead Byte Stuffing.
 *
 * Jumeau de `controller-2/Core/Src/comm/cobs.c`, y compris sur le cas limite du groupe plein
 * qui tombe exactement en fin de données : aucun octet de code supplémentaire n'est émis.
 * Les deux formes se décodent identiquement, mais les deux implémentations doivent produire
 * les mêmes octets — sinon la divergence se révèle le jour où une trame fait pile 254 octets
 * sans zéro. Les vecteurs partagés (`docs/protocol-vectors.json`) verrouillent ce point.
 */

/** Taille encodée maximale pour `n` octets bruts, délimiteur non compris. */
export function cobsMaxEncoded(n: number): number {
  return n + Math.floor(n / 254) + 1;
}

/** Encode sans écrire le délimiteur final : c'est l'appelant qui l'ajoute. */
export function cobsEncode(src: Uint8Array): Uint8Array {
  // +1 : le code final peut tomber hors de la longueur rendue (groupe plein en fin).
  const dst = new Uint8Array(cobsMaxEncoded(src.length) + 1);

  let out = 1; // la place du premier octet de code est réservée d'emblée
  let codePos = 0;
  let code = 1;

  for (let i = 0; i < src.length; i++) {
    const b = src[i]!;
    if (b !== 0) {
      dst[out++] = b;
      code++;
    }
    if (b === 0 || code === 0xff) {
      dst[codePos] = code;
      code = 1;
      codePos = out;
      if (b === 0 || i + 1 < src.length) {
        out++;
      }
    }
  }

  dst[codePos] = code;
  return dst.subarray(0, out);
}

/**
 * Décode une trame, délimiteur exclu. `src` ne doit contenir aucun 0x00.
 * Rend `null` si l'entrée est malformée — jamais une trame partielle : une trame douteuse
 * se jette, elle ne se répare pas.
 */
export function cobsDecode(src: Uint8Array): Uint8Array | null {
  if (src.length === 0) return null;

  const out: number[] = [];
  let i = 0;

  while (i < src.length) {
    const code = src[i]!;
    // 0x00 est le délimiteur : il ne doit jamais parvenir ici.
    if (code === 0) return null;
    // Le groupe annonce code-1 octets littéraux : ils doivent tenir dans l'entrée.
    if (i + code > src.length) return null;
    i++;

    for (let k = 1; k < code; k++) {
      out.push(src[i++]!);
    }

    // Zéro implicite en fin de groupe, sauf groupe plein et sauf en toute fin d'entrée.
    if (code !== 0xff && i < src.length) {
      out.push(0);
    }
  }

  return Uint8Array.from(out);
}
